exp/internal/controllers/machinepool_controller.go

```diff
@@ -253,33 +253,52 @@ func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv
 	return res, kerrors.NewAggregate(errs)
 }
 
-func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *expv1.MachinePool) error {
-	if ok, err := r.reconcileDeleteExternal(ctx, mp); !ok || err != nil {
-		// Return early and don't remove the finalizer if we got an error or
-		// the external reconciliation deletion isn't ready.
-		return err
-	}
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinepool *expv1.MachinePool) error {
+	deleteAllowed, clusterClient, err := r.isDeleteMachinePoolAllowed(ctx, cluster)
+
+	// Check for cluster allowing delete or machinePool delete timeout.
+	if deleteAllowed || r.isMachinePoolDeleteTimeoutPassed(machinepool) {
+		if ok, err := r.reconcileDeleteExternal(ctx, machinepool); !ok || err != nil {
+			// Return early and don't remove the finalizer if we got an error or
+			// the external reconciliation deletion isn't ready.
+			return err
+		}
 
-	if err := r.reconcileDeleteNodes(ctx, cluster, mp); err != nil {
-		// Return early and don't remove the finalizer if we got an error.
-		return err
+		// Delete nodes when cluster accessor available & there are nodes to delete.
+		if len(machinepool.Status.NodeRefs) > 0 && clusterClient != nil {
+			if err = r.deleteRetiredNodes(ctx, clusterClient, machinepool.Status.NodeRefs, machinepool.Spec.ProviderIDList); err != nil {
+				return err
+			}
+		}
+		// Remove finalizer to delete the resource.
+		controllerutil.RemoveFinalizer(machinepool, expv1.MachinePoolFinalizer)
+
+		return nil
 	}
 
-	controllerutil.RemoveFinalizer(mp, expv1.MachinePoolFinalizer)
-	return nil
+	return err
 }
 
-func (r *MachinePoolReconciler) reconcileDeleteNodes(ctx context.Context, cluster *clusterv1.Cluster, machinepool *expv1.MachinePool) error {
-	if len(machinepool.Status.NodeRefs) == 0 {
-		return nil
+// isMachinePoolDeleteTimeoutPassed check the machinePool node delete time out.
+func (r *MachinePoolReconciler) isMachinePoolDeleteTimeoutPassed(machinepool *expv1.MachinePool) bool {
```

Comment:
>// isMachinePoolDeleteTimeoutPassed check the machinePool node delete time out.

should this be named then `isNodeTimeoutPassed`? so we ref the specific "Node" time out and don't include the machinepool which implicit in the receiver

Reply:
Then I think it should isMachinePoolNodesTimeoutPassed or similar. Current naming implies a machinepool deletion timeout which there isn't one?

Reply:
okay